src/flower/grpc_server/flower_service_servicer.py

```diff
@@ -12,25 +12,94 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
-"""Servicer for FlowerService."""
-from typing import Iterator
+"""Servicer for FlowerService.
+
+Relevant knowledge for reading this modules code:
+    - https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+"""
+from typing import Callable, Iterator
 
 import grpc
 
 from flower.client_manager import ClientManager
+from flower.grpc_server.grpc_bridge import GRPCBridge
+from flower.grpc_server.grpc_proxy_client import GRPCProxyClient
 from flower.proto import transport_pb2_grpc
-from flower.proto.transport_pb2 import ClientRequest, ServerResponse
+from flower.proto.transport_pb2 import ClientMessage, ServerMessage
+
+
+def default_bridge_factory() -> GRPCBridge:
+    """Return NetworkClient instance."""
+    return GRPCBridge()
+
+
+def default_grpc_client_factory(cid: str, bridge: GRPCBridge) -> GRPCProxyClient:
+    """Return NetworkClient instance."""
+    return GRPCProxyClient(cid=cid, info={}, bridge=bridge)
+
+
+def register_client(
+    client_manager: ClientManager,
+    client: GRPCProxyClient,
+    context: grpc.ServicerContext,
+) -> bool:
+    """Try registering NetworkClient with ClientManager.
+    If not successful raise Exception."""
+    is_success = client_manager.register(client)
+
+    if is_success:
+
+        def rpc_termination_callback():
+            client.bridge.close()
+            client_manager.unregister(client)
+
+        context.add_callback(rpc_termination_callback)
+
+    return is_success
 
 
 class FlowerServiceServicer(transport_pb2_grpc.FlowerServiceServicer):
     """FlowerServiceServicer for bi-directional gRPC instruction stream."""
 
-    def __init__(self, client_manager: ClientManager) -> None:
-        self.client_manager = client_manager
+    def __init__(
+        self,
+        client_manager: ClientManager,
+        grpc_bridge_factory: Callable[[], GRPCBridge] = default_bridge_factory,
+        grpc_client_factory: Callable[
+            [str, GRPCBridge], GRPCProxyClient
+        ] = default_grpc_client_factory,
+    ) -> None:
+        self.client_manager: ClientManager = client_manager
+        self.grpc_bridge_factory = grpc_bridge_factory
+        self.client_factory = grpc_client_factory
 
     def Join(  # pylint: disable=invalid-name
-        self, request_iterator: Iterator[ClientRequest], context: grpc.ServicerContext
-    ) -> Iterator[ServerResponse]:
-        for _ in request_iterator:
-            # Yield empty message
-            yield ServerResponse()
+        self, request_iterator: Iterator[ClientMessage], context: grpc.ServicerContext,
+    ) -> Iterator[ServerMessage]:
+        """Method will be invoked by each NetworkClient which participates in the network.
+
+        Protocol:
+            - The first ClientMessage has always have the connect field set
```

Comment:
Does not reflect the updated protocol, instead:
- The first message is sent from the server to the client
- Both ServerMessage and ClientMessage are message "wrappers" wrapping the actual message
- The `Join` method is (pretty much) protocol unaware